### PR TITLE
Fix interpolation on stretched grid

### LINKED ISSUE #2395: Adding Lagrangian particles to a nonuniform grid
Hi Oceananigans Community,

As discussed with @glwagner and @johnryantaylor, I post an issue about adding Lagrangian particles to a nonuniform grid in the z-direction. Hope the following minimum working example is mini enough and bug-free. Happy to learn and discuss. 

Best,
Si Chen

<details open>
<summary>Minimum Working Example</summary>

```
using Random
using Printf
using Plots

using Oceananigans
using Oceananigans.Units: minute, minutes, hour

Nz = 24          # number of points in the vertical direction
Lz = 32          # (m) domain depth

refinement = 1.2 # controls spacing near surface (higher means finer spaced)
stretching = 12  # controls rate of stretching at bottom

h(k) = (k - 1) / Nz

ζ₀(k) = 1 + (h(k) - 1) / refinement

Σ(k) = (1 - exp(-stretching * h(k))) / (1 - exp(-stretching))

z_faces(k) = Lz * (ζ₀(k) * Σ(k) - 1)

grid = RectilinearGrid(size = (32, 32, Nz), 
                          x = (0, 64),
                          y = (0, 64),
                          #z = (-Lz, 0))         # Errors are gone for uniform grid 
                          z = z_faces)

############################ Lagrangian Particles
n_particles = 10;
x₀ = 60*rand(n_particles);   
y₀ = 60*rand(n_particles);
z₀ = -Lz*0.5*rand(n_particles);  
lagrangian_particles = LagrangianParticles(x=x₀, y=y₀, z=z₀, restitution=0)
############################

b_bcs = FieldBoundaryConditions(top=FluxBoundaryCondition(1e-8))

model = NonhydrostaticModel(advection = UpwindBiasedFifthOrder(),
                            timestepper = :RungeKutta3,
                            grid = grid,
                            tracers = (:b,),
                            coriolis = FPlane(f=1e-4),
                            buoyancy = BuoyancyTracer(),
                            closure = AnisotropicMinimumDissipation(),
                            boundary_conditions = (b=b_bcs,),
                            particles=lagrangian_particles) ############################ Lagrangian Particles


bᵢ(x, y, z) = 1e-5 * z + 1e-9 * rand()

set!(model, b=bᵢ)


simulation = Simulation(model, Δt=10.0, stop_time=20minutes)

wizard = TimeStepWizard(cfl=1.0, max_change=1.1, max_Δt=1minute)

simulation.callbacks[:wizard] = Callback(wizard, IterationInterval(10))


run!(simulation)
```
</details>

### PR DESCRIPTION
**Pull request by simone-silvestri**
Closes #2395

### CODE REVIEW

File: src/Fields/interpolate.jl [hunk 1]
```diff
@@ -1,16 +1,16 @@
...
 
-    @CUDA.allowscalar begin
-        y₂ = searchsortedfirst(vec, val)
-        y₁ = searchsortedlast(vec,  val)
+    allowscalar(true)
+    y2 = searchsortedfirst(vec, val)
+    y1 = searchsortedlast(vec,  val)
+    x2 = vec[y2]
+    x1 = vec[y1]
+    allowscalar(false)
```
Comment (glwagner) [inline src/Fields/interpolate.jl:11]: 🤦‍♂️

Comment (glwagner) [reply src/Fields/interpolate.jl:11]: Is this insanely slow on the GPU?

File: src/Fields/interpolate.jl [hunk 2]
```diff
@@ -1,16 +1,16 @@
...
+using CUDA: allowscalar 
 
 @inline function linear_interpolate_sorted_vector(vec, val)
 
-    @CUDA.allowscalar begin
-        y₂ = searchsortedfirst(vec, val)
-        y₁ = searchsortedlast(vec,  val)
+    allowscalar(true)
+    y2 = searchsortedfirst(vec, val)
+    y1 = searchsortedlast(vec,  val)
```
Comment (glwagner) [inline src/Fields/interpolate.jl:8]:
```suggestion
    y1 = searchsortedlast(vec, val)
```

**Comment by glwagner** [comment]
need a test too I think. Probably some refactoring to come for Lagrangian particles...

File: src/Fields/interpolate.jl (hunk 1 above)
Comment (simone-silvestri) [reply src/Fields/interpolate.jl:11]: for sure it is, not in case of `LagrangianParticles` because the `interpolate` function is called from a kernel

**Review by glwagner — APPROVED** [review]

**Comment by glwagner** [comment]
Test can come now or later

**Comment by navidcy** [comment]
@simone-silvestri, merge main and resolve conflicts?

**Comment by tomchor** [comment]
This seems reasonably important. Any particular reason why this PR isn't being merged? Has it been superseded by another?

**Pull request closed without merging**